### PR TITLE
#164797147 Ft Enable twitter authentication

### PR DESCRIPTION
**Pull request by musonant**
#### What does this PR do?
* Setup Twitter authentication

#### Description of Task to be completed?
* Write tests to test twitter authentication
* Setup twitter with twitter-passport

#### How should this be manually tested?
* After cloning repo and starting server
* On your browser, attempt to open `http:localhost:3000/api/v1/auth/twitter`

#### Any background context you want to provide?
* NA

#### What are the relevant pivotal tracker stories?(if applicable)
#164797147

#### Screenshots
Screenshot 2019-04-11 at 6 12 44 PM
Screenshot 2019-04-11 at 6 10 03 PM
Screenshot 2019-04-11 at 6 09 56 PM
Screenshot 2019-04-11 at 6 14 35 PM

### CODE REVIEW

File: src/db/migrations/20190325172457-create-users.js [hunk 1]
```diff
@@ -29,7 +29,7 @@ export default {
       },
       password: {
         type: Sequelize.STRING,
-        allowNull: false,
+        allowNull: true,
```
Comment (nwamugo) [inline src/db/migrations/20190325172457-create-users.js:32]: I see we are allowing the possibility of a user to sign in, without a password. Could you kindly tell why we are doing so?

File: src/db/models/User.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ export default (sequelize, DataTypes) => {
       },
       email: {
         type: DataTypes.STRING,
+        allowNull: true,
```
Comment (nwamugo) [inline src/db/models/User.js:30]: How does this affect the other non-social way of signing up?

File: src/routes/auth.route.js [hunk 3]
```diff
@@ -98,6 +98,87 @@ authRouter.get(
...
+ *        description: Unexpected error.
+ */
+authRouter.get('/twitter', passport.authenticate('twitter'));
+authRouter.get(
+  '/twitter/callback',
+  passport.authenticate('twitter', { session: false }),
+  socialController,
+);
+
+/**
```
Comment (nwamugo) [inline src/routes/auth.route.js]: I notice this is exactly the same endpoint as above. Is the duplication intentional?

File: src/routes/auth.route.js [hunk 4]
```diff
@@ -98,6 +98,87 @@ authRouter.get(
...
```
Comment (nwamugo) [inline src/routes/auth.route.js]: Likewise here from lines 155 to 180. Meaning we have three exact identical endpoints in this file. Is that how it is designed to work?

File: src/routes/services/passport-strategies.services.js [hunk 5]
```diff
@@ -1,12 +1,13 @@
 import passport from 'passport';
 import dotenv from 'dotenv';
+import { Strategy as FacebookStrategy } from 'passport-facebook';
+import { Strategy as TwitterStrategy } from 'passport-twitter';
```
Comment (nwamugo) [inline src/routes/services/passport-strategies.services.js:4]: Nice import naming

File: src/routes/services/passport-strategies.services.js [hunk 6]
```diff
@@ -35,6 +49,25 @@ const facebookAuth = async (accessToken, refreshToken, profile, done) => {
...
+      where: { socialId: profile.id, socialProvider: 'twitter' },
+      defaults: {
+        firstName: profile.username,
+        username: profile.emails[0].value,
+        email: profile.emails[0].value,
+        socialProvider: profile.provider,
+      },
+    });
+
+    return done(null, user);
```
Comment (nwamugo) [inline src/routes/services/passport-strategies.services.js:64]: I observe you are using both 'async await' and 'done' in this function. Are they doing different things?

File: src/tests/routes/controllers/social-auth.test.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        expect(response.body.status).to.eql('success');
+        expect(response.body.data.user).to.be.an('object');
+        expect(response.body.data.user.firstName).to.eql(
+          req.user.dataValues.firstName,
+        );
+        expect(response.body.data.token).to.be.a('string');
+      });
+    });
+  });
+});
```
Comment (nwamugo) [inline src/tests/routes/controllers/social-auth.test.js:121]: Fine effort in your testing. Well done!

File: src/tests/routes/controllers/social-auth.test.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        provider: 'twitter',
+      };
+      twitterTestUser = await twitterAuth(token, tokenSecret, profile, done);
+      expect(twitterTestUser.dataValues.id).to.be.a('string');
+      expect(twitterTestUser.dataValues.email).to.eql('testMail.tw.com');
+      expect(twitterTestUser.dataValues.username).to.eql('testMail.tw.com');
+    });
+  });
+  describe('Social authentication controller', () => {
+    describe('should process user received from passport authentication', async () => {
```
Comment (igbominadeveloper) [inline src/tests/routes/controllers/social-auth.test.js:64]: I think you should substitute `it` for `describe` here to make the test suite readable and understandable.

File: src/tests/routes/controllers/social-auth.test.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        status(code) {
+          res.statusCode = code;
+          return res;
+        },
+        json(data) {
+          return { body: data, status: res.statusCode };
+        },
+      };
+
+      it('and return the appropriate response for a new user', async () => {
```
Comment (igbominadeveloper) [inline src/tests/routes/controllers/social-auth.test.js:99]: could you look into the sentence construction here to make it more readable?

File: src/tests/routes/controllers/social-auth.test.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        req.user._options.isNewRecord = true;
+        const response = await socialController(req, res);
+
+        expect(response.status).to.eql(201);
+        expect(response.body.data.token).to.be.a('string');
+      });
+
+      it('and return the appropriate response for an existing user', async () => {
```
Comment (igbominadeveloper) [inline src/tests/routes/controllers/social-auth.test.js:107]: And here too...

File: src/routes/auth.route.js (hunk 3 above)
Comment (nwamugo) [inline src/routes/auth.route.js]: I notice this is exactly the same endpoint as line 101 to line 125. Is the duplication intentional?

File: src/db/migrations/20190325172457-create-users.js (hunk 1 above)
Comment (musonant) [reply src/db/migrations/20190325172457-create-users.js:32]: When a user registers with a third-party application like Facebook, his password would not be provided by Facebook. However, he is able to create/edit password on his profile.

File: src/db/models/User.js (hunk 2 above)
Comment (musonant) [reply src/db/models/User.js:30]: Those methods are validated, which means that signup operations will not be successful unless an email is provided.

File: src/tests/routes/controllers/social-auth.test.js (hunk 7 above)
Comment (musonant) [reply src/tests/routes/controllers/social-auth.test.js:121]: Oh, thanks.

File: src/routes/auth.route.js (hunk 4 above)
Comment (musonant) [reply src/routes/auth.route.js]: Oh yeah! I'm glad you spotted this. I will fix it in a bit.
Thanks.

File: src/routes/services/passport-strategies.services.js (hunk 6 above)
Comment (musonant) [reply src/routes/services/passport-strategies.services.js:64]: `Done` is used by passportJs during authentication.

File: src/tests/routes/controllers/social-auth.test.js (hunk 9 above)
Comment (musonant) [reply src/tests/routes/controllers/social-auth.test.js:99]: This statement continues from the describe block at the top of it. When running tests, it shows up like this:
Social controller should process user received from passport authentication
    -  and return the appropriate response for a new user

File: src/routes/services/passport-strategies.services.js (hunk 6 above)
Comment (nwamugo) [reply src/routes/services/passport-strategies.services.js:64]: Okay. My thoughts are that `done` and `async await` do the same thing. Could we try using done alone without the async await? Or async await without the done? Let us see if any of the substitution will work.

File: src/tests/routes/controllers/social-auth.test.js (hunk 9 above)
Comment (emmsdan) [reply src/tests/routes/controllers/social-auth.test.js:99]: Nice

File: package.json [hunk 11]
```diff
@@ -55,10 +55,12 @@
   "author": "Andela Simulations Programme",
   "license": "MIT",
   "dependencies": {
+    "axios": "^0.18.0",
```
Comment (Japhethca) [inline package.json]: We can remove this dependency since we are not using it.

File: .travis.yml [hunk 12]
```diff
@@ -1,9 +1,9 @@
 language: node_js
 node_js:
-  - "stable"
+  - '11.11.0'
```
Comment (Japhethca) [inline .travis.yml]: we can use an LTS node version as it's more stable.

**Pull request closed without merging**